developer/src/kmc-analyze/src/osk-character-use/index.ts

```diff
@@ -0,0 +1,265 @@
+    const source = reader.read(this.callbacks.loadFile(filename));
+    // TODO: handle errors
+    const scanKey = (key: TouchLayout.TouchLayoutKey | TouchLayout.TouchLayoutSubKey) => {
+      if(!key.text) {
+        return;
+      }
+      if(key.text.length > 2 && key.text[0] == '*' && key.text[key.text.length-1] == '*') {
+        // Don't add '*special*' key captions
+        return;
+      }
+      strings.push(key.text);
+    }
+
+    const scanPlatform = (platform: TouchLayout.TouchLayoutPlatform) => {
+      if(!platform) {
+        return;
+      }
+      for(let layer of platform.layer) {
+        for(let row of layer.row) {
+          for(let key of row.key) {
+            scanKey(key);
+            let f: keyof TouchLayout.TouchLayoutFlick;
+            for(f in key.flick ?? {}) {
+              scanKey(key.flick[f]);
+            }
+            for(let sk of key.sk ?? []) {
+              scanKey(sk);
+            }
+            for(let mt of key.multitap ?? []) {
+              scanKey(mt);
+            }
+          }
+        }
+      }
+    }
+    scanPlatform(source.desktop);
+    scanPlatform(source.phone);
+    scanPlatform(source.tablet);
+    return strings;
+  }
+
+  //
+  // Results reporting
+  //
+
+  public get strings(): string[] {
+    // Sort and remove duplicates
+    const strings = [...new Set(this._strings)];
+    strings.sort();
+    return strings;
+  }
+
+  public getStrings(format?: 'text'|'markdown'|'json'): string[] {
+    switch(format) {
+      case 'markdown':
+        return AnalyzeOskCharacterUse.getStringsAsMarkdown(this.strings);
+      case 'json':
+        return AnalyzeOskCharacterUse.getStringsAsJson(this.strings);
+      }
+    return AnalyzeOskCharacterUse.getStringsAsText(this.strings);
+  }
+
+  // Following functions are static so that we can keep them pure
+  // and potentially refactor into separate reporting class later
+
+  private static getStringsAsText(strings: string[]) {
+    let lines: string[] = [];
+    for(let s of strings) {
+      let ux = this.stringToUnicodeSequence(s);
+      lines.push(ux + '\t' + s);
+    }
+    return lines;
+  }
+
+  private static getStringsAsMarkdown(strings: string[]) {
+    let lines: string[] = [];
+    lines.push('Code Points | Key Caps');
+    lines.push('------------|---------');
+    for(let s of strings) {
+      let ux = this.stringToUnicodeSequence(s);
+      lines.push(ux + ' | ' + this.escapeMarkdownChar(s));
+    }
+    return lines;
+  }
+
+  private static getStringsAsJson(strings: string[]) {
+    let data: string[] = [];
+    for(let s of strings) {
+      data.push(s);
+    }
+    return JSON.stringify(data, null, 2).split('\n');
+  }
+
+  private static escapeMarkdownChar(s: string) {
+    // commonmark 2.4: all punct can be escaped
+    const punct = '!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~';
+
+    if(punct.includes(s)) {
+      return '\\' + s;
+    }
+
+    // TODO: there are other chars, use a library or regex match?
```

Comment:
Yeah, that gets 'fun' depending on Markdown flavor:

https://michelf.ca/projects/php-markdown/extra/#backslash

> Markdown Extra adds the colon (:) and the pipe (|) to the list of characters you can escape using a backslash. With this you can prevent them from triggering a definition list or a table.

Noting this one in particular since it's what we use for help.keyman.com.

Reply:
```suggestion
    // TODO: there are other chars, use a library or regex match? e.g. :, | https://michelf.ca/projects/php-markdown/extra/#backslash
```